JAVA-1193 + JAVA-1120: Refresh token and replica metadata synchronously when schema is altered.

adutra opened the pull request:
I'm also including a fix for JAVA-1120 in the same PR as the affected code is pretty much the same.

Note that I don't have specific tests for JAVA-1120 but I did benchmark quickly the creation of 50 tables with default debouncer options. Without the fix it takes in avg. 50 secs on my laptop; with the fix, it takes 3 secs in avg.

The 3rd commit brings minor enhancements to `SchemaChangesTest`: missing assertions and verifications basically.

Commit:
JAVA-1193: Refresh token and replica metadata synchronously when schema is altered

This commit also introduces the following refactorings:

1) Cache and reuse host and token information whenever possible

This commit modifies the TokenMap class so that it caches all
the information about hosts and tokens from the last token map
computation. This information can then be reused in a subsequent
token map computation, if only the schema has changed, but not
the topology.

2) Optimize calls to submitNodeListRefresh()

In some circonstances, this method could be called recursively
(i.e., triggered while already processing a node list refresh request).

This commit suppresses the calls to submitNodeListRefresh() from
ControlConnection and replaces them with calls to that method only
when appropriate, i.e.:
- When inside a node refresh, and the node has been added or removed;
- When reconnecting, and the node has been added.

The effect is that submitNodeListRefresh() is NOT called when in
triggerOnAdd() nor in triggerOnRemove(), because we are already inside
a node list refresh operation.

3) Minor optimizations

This commit optimizes code around TokenMap:
- Remove unnecessary field TokenMap.hosts
- Rename fields in TokenMap for consistency
- Avoid creating intermediary Set<String> instances by creating
Set<Token> instances directly when reading rows from system.local and
system.peers.

adutra commented on driver-core/src/main/java/com/datastax/driver/core/Cluster.java:
```diff
@@ -2265,7 +2267,20 @@ public void run() {
+                        if (refreshSchema) {
+                            schemaReady = submitSchemaRefresh(targetType, targetKeyspace, targetName, targetSignature);
+                            if (refreshNodeList) {
+                                schemaReady = Futures.transform(schemaReady, new AsyncFunction<Object, Void>() {
```
Note that a call to `submitNodeListRefresh()` is already done in `ControlConnection.refreshSchema()` for other cases (notably when a schema change event is received), but here we need to wait on that future. I think the cost of enqueueing 2 node list refresh requests is negligible compared to the advantage of having the token metadata guaranteed to be up-to-date.

Commit:
JAVA-1120: Skip schema refresh debouncer when checking for agreement as a result of schema change made by client

This commit contains substantial contributions by Andrew Tolbert (@tolbertam).

This commit removes nonDebouncingQueryOptions from tests that no longer require it.

adutra commented on driver-core/src/main/java/com/datastax/driver/core/Cluster.java:
```diff
@@ -2265,7 +2267,25 @@ public void run() {
+                                schemaReady = Futures.transform(schemaReady, new AsyncFunction<Object, Void>() {
+                                    @Override
+                                    public ListenableFuture<Void> apply(Object input) throws Exception {
+                                        ListenableFuture<Void> nodeListRefreshReady = submitNodeListRefresh();
```
Note that a call to `submitNodeListRefresh()` is already done in `ControlConnection.refreshSchema()` for other cases (notably when a schema change event is received), but here we need to wait on that future. I think the cost of enqueueing 2 node list refresh requests is negligible compared to the advantage of having the token metadata guaranteed to be up-to-date.

adutra commented:
Added two new commits:

1. 1de87db suppresses recursive call to `submitNodeListRefresh()`
2. b7e9635 optimizes token metadata computation during schema resfreshes (and makes it synchronous)

Please review those carefully as they affect sensitive areas, and I might have let slip something important.

adutra commented:
Rebased and squashed on top of current 3.0.x.

tolbertam commented:
We can probably limit the usage of `TestUtils.nonDebouncingQueryOptions` in some cases with JAVA-1120.  I'll work on that, would be a good way of ensuring this fix, as these options should no longer be needed in the schema change case.

tolbertam commented:
Not sure why appveyor triggered a check here..please ignore that.

adutra commented:
Looks like CI is happy with the removal of `TestUtils.nonDebouncingQueryOptions`, and more importantly, the overall build time remained constant.

olim7t commented on driver-core/src/main/java/com/datastax/driver/core/QueryOptions.java:
```diff
@@ -299,7 +301,19 @@ public QueryOptions setMetadataEnabled(boolean enabled) {
+                    // this time with information about keyspaces
+                    manager.submitSchemaRefresh(null, null, null, null);
+                }
+            });
```
Do we need to refresh the node list here? `metadataEnabled == false` doesn't disable the handling of topology events, so the node list should be up to date.

adutra replied:
The hosts list is kept up-to-date, but not the token map, which is likely to be `null` at this point. The field `Metadata.tokenMap` is only ever updated if `metadataEnabled == true` (see [here](https://github.com/datastax/java-driver/blob/3.0.x/driver-core/src/main/java/com/datastax/driver/core/ControlConnection.java#L656-L657)), and the only way to build the token map from scratch for the first time is to trigger a node list refresh.

olim7t commented on driver-core/src/main/java/com/datastax/driver/core/Metadata.java:
```diff
@@ -633,20 +660,20 @@ private TokenMap(Token.Factory factory,
 
             Set<Host> hosts = allTokens.keySet();
             Map<Token, Host> tokenToPrimary = new HashMap<Token, Host>();
             Map<Host, Set<Token>> primaryToTokens = new HashMap<Host, Set<Token>>();
```
Now that the string -> token conversion is handled upstream, this is just a deep copy of `allTokens`.

olim7t commented on driver-core/src/main/java/com/datastax/driver/core/Metadata.java:
```diff
@@ -633,20 +660,20 @@ private TokenMap(Token.Factory factory,
         }
 
-        public static TokenMap build(Token.Factory factory, Map<Host, Collection<String>> allTokens, Collection<KeyspaceMetadata> keyspaces) {
+        public static TokenMap build(Token.Factory factory, Map<Host, ? extends Collection<Token>> allTokens, Collection<KeyspaceMetadata> keyspaces) {
```
I like the idea of a simplified computation when only the keyspaces have changed (not the node list). But there's still a lot of keyspace-independent stuff that's recomputed here, basically everything except `hostsToRanges` (even `replStrategyToHosts` would be nice to cache between invocations). I wonder if we should not cache more aggressively to avoid this.

adutra commented:
Rebased and addressed feedback: the `TokenMap` class now caches all the information required to rebuild itself when only the schema has changed, but not the topology.

tolbertam commented on driver-core/src/main/java/com/datastax/driver/core/Metadata.java:
```diff
@@ -66,34 +64,43 @@
-        Token.Factory factory = getTokenFactory(partitioner);
-        if (factory == null) return;
-        rebuildTokenMap(factory, tokenMap.primaryToTokens);
+        lock.lock();
```
Very nice!

adutra replied:
@tolbertam FYI I added another commit with further optimizations around TokenMap re-creation.

olim7t commented:
Looks good to me. We cache more info than before, but it's O(numTokens) so it should still be reasonable, even for large clusters.
@adutra could you squash the commits as appropriate, and reword the messages? (no "should be" or "consider...", just imperative sentences)

adutra commented on driver-core/src/main/java/com/datastax/driver/core/SchemaParser.java:
```diff
@@ -63,6 +63,9 @@ void refresh(Cluster cluster,
                 updateKeyspaces(metadata, metadata.keyspaces, keyspaces, targetKeyspace);
+                // If we rebuild all from scratch or have an updated keyspace, rebuild the token map
+                // since some replication on some keyspace may have changed
+                metadata.rebuildTokenMap();
```
I am still a bit worried bout this call being synchronous, but not atomic. Here, right before the call to `rebuildTokenMap`, the field `Metadata.keyspaces` would contain the changes, but not the field `Metadata.tokenMap`. There is still a slight, albeit improbable, chance that an observer could retrieve inconsistent data. What makes it difficult to fix is that updates to `Metadata.keyspaces` are made on-the-go.

olim7t replied:
> I am still a bit worried bout this call being synchronous, but not atomic

I agree, but fixing that is much more involved: I think the best way would be to make `Metadata` an immutable object that contains a snapshot of _all_ metadata at a given point in time. After a single call to `cluster.getMetadata()` the user would have the guarantee that all fields are consistent with each other (which is not possible currently since you call methods on a mutable object).
Internally we would handle updates by storing this object in a volatile field, and switching it with a new copy when something gets updated.
This is likely to break the API though.
